### ISSUE TITLE
Allow setting the default value of an optional parameter to one of the non-default parameters

### ISSUE THREAD

**Issue by Bretttt** [dotnet/roslyn#13595]
**Version Used**: 

C# 5

**Steps to Reproduce**:

When you are declaring a method with optional parameters, allow the ability to set the default value to one of the previous, non-optional parameters. This feature would help simplify many cases by reducing the need for overloads, and even allow for functionality that was not previously possible. Consider the following example:

```
public class FooClass
{
    private int _foo, _bar, _baz, _qux;

    public FooClass(int foo, int bar, int baz = foo, int qux = bar / 2)
    {
        _foo = foo;
        _bar = bar;
        _baz = baz;
        _qux = qux;
    }
```

There is currently no way to gain this exact functionality. This example illustrates my point:

```
public class FooClass
{
    private int _foo, _bar, _baz, _qux;

    public FooClass(int foo, int bar, int baz, int qux)
    {
        _foo = foo;
        _bar = bar;
        _baz = baz;
        _qux = qux;
    }

    public FooClass(int foo, int bar)
    {
        _foo = foo;
        _bar = bar;
        _baz = foo;
        _qux = bar / 2;
    }

    public FooClass(int foo, int bar, int baz)
    {
        _foo = foo;
        _bar = bar;
        _baz = baz;
        _qux = bar / 2;
    }

    public FooClass(int foo, int bar, int qux)
    {
        _foo = foo;
        _bar = bar;
        _baz = foo;
        _qux = quz;
    }
```

Obviously, a call to one of the three-argument constructors would result in an error because the argument types are the same, making them indistinguishable.

Here is an example of overload reduction. This:

```
public static void Foo(int arg0, int arg1)
{
    //...
}

public static void Foo(int arg0)
{
    Foo(arg0, arg1);
}
```

Could be reduced to this:

```
public static void Foo(int arg0, int arg1 = arg0)
{
    //...
}
```

**Expected Behavior**:

The ability to set default values of optional parameters to expressions including non-optional arguments.

**Actual Behavior**:

> error CS0103: The name 'foo' does not exist in the current context
> error CS0103: The name 'bar' does not exist in the current context

**Comment by HaloFour** [comment]
Optional parameters are based on a CLR feature that requires that the default value be a constant that can be directly embedded in the BLOB stream of the assembly.  There's no mechanism for C# to embed an expression or code as the default value.

**Comment by gafter** [comment]
@HaloFour We already use attributes to convey this kind of information, e.g. `DateTime` and `Decimal` default parameter values. We could extend the set of attributes we use to convey this from the compiler to itself.

**Comment by HaloFour** [comment]
@gafter 

Of course, but I'm going to make the assumption that attempting to support arbitrary expressions as the default value for a parameter is probably a bit much to encode within attributes.  The proposal above not only refers to other arguments within the same method call but also performs arbitrary operations against those arguments.  I'd argue that it could quickly become a rabbit hole and you'd end up defining a language within a language.

As for just supporting default values referencing other parameters, that would probably be relatively simple, but I'd like to see a use case that isn't just contrived.  I could see adding support for other types as well, e.g. `CancellationToken cancellationToken = CancellationToken.None`, which reads and self-documents better than `CancellationToken cancellationToken = default(CancellationToken)`.

**Comment by paulomorgado** [comment]
@Bretttt, what do you have against overloads?

**Comment by CyrusNajmabadi** [comment]
Closing this out.  We're doing all language design now at dotnet/csharplang.  If you're still interested in this idea let us know and we can migrate this over to a discussion in that repo.  Thanks!
